File: SampleApp/app/src/main/java/jp/fluct/sample/samplefluctsdkapp/ContentsListFragment.java
package jp.fluct.sample.samplefluctsdkapp;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import jp.fluct.fluctsdk.FluctAdBanner;

public class ContentsListFragment extends Fragment {

    private ListView mContentListView;

    public ContentsListFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_contents_list, container, false);
        mContentListView = (ListView) layout.findViewById(R.id.list);

        return layout;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        List<Object> listContents = generateContents();
        ContentsListAdapter adapter = new ContentsListAdapter(getContext(), listContents);

        // Fluct広告を入れる
        FluctAdBanner.SizeType sizeType = FluctAdBanner.SizeType.BANNER;
        String mediaId = "0000005617";
        adapter.insert(new Ad(sizeType, mediaId), 3);
        adapter.insert(new Ad(sizeType, mediaId), 12);
        adapter.insert(new Ad(sizeType, mediaId), 30);
        adapter.insert(new Ad(sizeType, mediaId), 55);
        adapter.insert(new Ad(sizeType, mediaId), 80);

        // リストへ反映
        mContentListView.setAdapter(adapter);
    }

    // リストコンテンツの作成
    private List<Object> generateContents(){
        List<Object> contents = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            Content content = new Content(i);
            content.name = "item " + (i + 1);
            contents.add(content);
        }
        return contents;
    }

    /**
     * コンテンツのモデル
     */
    private static class Content {
        final int identity;
        String name;

        Content(int id) {
            this.identity = id;
        }
    }

    /**
     * 広告モデル
     */
    private static class Ad {
        final FluctAdBanner.SizeType sizeType;
        final String mediaId;

        Ad(FluctAdBanner.SizeType sizeType, String mediaId) {
            this.sizeType = sizeType;
            this.mediaId = mediaId;
        }
    }

    private static class ContentsListAdapter extends ArrayAdapter<Object> {

        private final LayoutInflater mInflater;

        public ContentsListAdapter(Context context, List<Object> contents) {
            super(context, R.layout.item_list_content, contents);
            mInflater = LayoutInflater.from(context);
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            Object content = getItem(position);

            int itemViewType = getItemViewType(position);
            ViewType viewType = ViewType.find(itemViewType);
            switch (viewType) {
                case CONTENT:
                    convertView = getViewForContent((Content) content, position, convertView, parent);
                    break;
                case AD:
                    convertView = getViewForFluctAd((Ad) content, position, convertView, parent);
                    break;
            }

            return convertView;
        }

        private View getViewForContent(Content content, int position, View convertView, ViewGroup parent) {
            ContentViewHolder holder;
            if (convertView == null) {
                convertView = mInflater.inflate(R.layout.item_list_content, null);
                holder = new ContentViewHolder();
                holder.index = (TextView) convertView.findViewById(R.id.list_index);
                holder.name = (TextView) convertView.findViewById(R.id.list_name);
                holder.thumbnail = (ImageView) convertView.findViewById(R.id.list_thumbnail);
                convertView.setTag(holder);
            } else {
                holder = (ContentViewHolder) convertView.getTag();
            }

            // リスト番号
            holder.index.setText(String.valueOf(position + 1));

            // 項目名
            holder.name.setText(content.name);

            return convertView;
        }

        private View getViewForFluctAd(Ad ad, int position, View convertView, ViewGroup parent) {
            FluctAdBanner adView;
            if (convertView == null) {
                // 初回作成
                adView = new FluctAdBanner(getContext());
                adView.getViewSettings().setSizeType(ad.sizeType);
                adView.getViewSettings().setMediaId(ad.mediaId);
                adView.setTag(ad);

            } else {
                // 広告の再利用
                adView = (FluctAdBanner) convertView;
                Ad preListAd = (Ad) adView.getTag();

                if (preListAd != ad) {
                    // リサイクルされた広告と違う広告を表示させたい場合は再生成する
                    adView = new FluctAdBanner(getContext());
                    adView.getViewSettings().setSizeType(ad.sizeType);
                    adView.getViewSettings().setMediaId(ad.mediaId);
                    adView.setTag(ad);
                }
            }

            return adView;
        }


        @Override
        public int getViewTypeCount() {
            return ViewType.values().length;
        }

        @Override
        public int getItemViewType(int position) {
            Object content = getItem(position);
            if (content instanceof Ad) {
                // 広告
                return ViewType.AD.id;

            } else {
                // デフォルト
                return ViewType.CONTENT.id;

            }
        }

        private enum ViewType {
            CONTENT(0), AD(1);

            final int id;

            ViewType(int id) {
                this.id = id;
            }

            static ViewType find(int id) {
                for(ViewType vt: values()){
                    if(vt.id == id){
                        return vt;
                    }
                }
                return CONTENT;
            }
        }

        private static class ContentViewHolder {
            TextView index;
            TextView name;
            ImageView thumbnail;
        }
    }
}

File: SampleApp/kotlin/app/src/main/java/jp/fluct/sample/kotlinapp/ContentsListFragment.kt
package jp.fluct.sample.kotlinapp

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import jp.fluct.fluctsdk.FluctAdBanner
import java.util.*

class ContentsListFragment : Fragment() {

    private var contentsRecyclerView: RecyclerView? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val layout = inflater.inflate(R.layout.fragment_contents_list, container, false)
        contentsRecyclerView = layout.findViewById(R.id.content_list)

        return layout
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        val listContents: MutableList<ListModel> = generateContents() as MutableList<ListModel>
        val listView = view!!.findViewById<RecyclerView>(R.id.content_list)
        val adapter = ContentsListAdapter(activity!!.applicationContext, listContents)

        // Fluct広告を入れる
        val sizeType = FluctAdBanner.SizeType.BANNER
        val mediaId = "0000005617"
        adapter.insertItem(Ad(sizeType, mediaId), 3)
        adapter.insertItem(Ad(sizeType, mediaId), 12)
        adapter.insertItem(Ad(sizeType, mediaId), 30)
        adapter.insertItem(Ad(sizeType, mediaId), 55)
        adapter.insertItem(Ad(sizeType, mediaId), 80)

        listView.adapter = adapter
    }

    // リストコンテンツの作成
    private fun generateContents(): MutableList<Content> {
        val contents = ArrayList<Content>()
        for (i in 0..99) {
            val name = "item " + (i + 1)
            val content = Content(i, name)
            contents.add(content)
        }
        return contents
    }

    abstract class ListModel

    /**
     * コンテンツのモデル
     */
    class Content(val identity: Int, val name: String) : ListModel()

    /**
     * 広告モデル
     */
    class Ad(val sizeType: FluctAdBanner.SizeType, val mediaId: String) : ListModel()

    class ContentsListAdapter(
            val context: Context,
            private val itemList: MutableList<ListModel>
    ) : RecyclerView.Adapter<BaseViewHolder>() {

        fun insertItem(content: ListModel, index: Int) {
            this.itemList.add(index, content)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
            return when (viewType) {
                ViewType.CONTENT.ordinal -> {
                    val view = LayoutInflater.from(context).inflate(R.layout.item_list_content, parent, false)
                    ContentViewHolder(view)
                }
                ViewType.AD.ordinal -> {
                    val fluctAdBanner = FluctAdBanner(context)
                    AdViewHolder(fluctAdBanner)
                }
                else -> throw IllegalStateException()
            }
        }

        override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
            val item = itemList[position]
            when (item) {
                is Content -> {
                    val contentVH = holder as ContentViewHolder
                    val index = position + 1
                    contentVH.index?.text = "$index"
                    contentVH.name?.text = item.name
                }
                is Ad -> {
                    val bannerView = holder.itemView as FluctAdBanner
                    bannerView.viewSettings.mediaId = item.mediaId
                }
            }
        }

        override fun getItemCount(): Int {
            return this.itemList.size
        }

        override fun getItemViewType(position: Int): Int {
            val content = this.itemList[position]
            return if (content is Ad) ViewType.AD.ordinal else ViewType.CONTENT.ordinal
        }

        enum class ViewType {
            CONTENT, AD
        }
    }

    abstract class BaseViewHolder(view: View) : RecyclerView.ViewHolder(view)

    class AdViewHolder(view: View) : BaseViewHolder(view)

    class ContentViewHolder(view: View) : BaseViewHolder(view) {
        var index: TextView? = view.findViewById(R.id.list_index)
        var name: TextView? = view.findViewById(R.id.list_name)
    }
}